sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala

```diff
@@ -58,12 +58,12 @@ trait QueryExecutionListener {
    * @param funcName the name of the action that triggered this query.
    * @param qe the QueryExecution object that carries detail information like logical plan,
    *           physical plan, etc.
-   * @param exception the exception that failed this query.
+   * @param error the exception that failed this query.
    *
    * @note This can be invoked by multiple different threads.
    */
   @DeveloperApi
-  def onFailure(funcName: String, qe: QueryExecution, exception: Exception): Unit
+  def onFailure(funcName: String, qe: QueryExecution, error: Throwable): Unit
```

Comment:
This will need a release note.

Reply:
There is another issue in this API: It passes a private class `QueryExecution` to the user in a public API. Didn't fix it as it will require a re-design of this API.

Reply:
Yea, we should sort those problems out. There's another occurrence at Dataset.queryExecution as an API.

Reply:
I'm a bit concerned about this change. This broke one of my project that uses this API. Instead of changing the signature here, did we consider just wrapping the exception?

Reply:
RE: `QueryExecution` I think that has become a developer API at this point.  I think a lot of developers use it to debug things.

Reply:
> wrapping the exception?

This is a good suggestion. This can avoid api changes, and we can also fix this issue in 2.4.